File: aat-android/src/main/java/ch/bailu/aat/services/location/GpsLocation.java
package ch.bailu.aat.services.location;

import android.content.Context;
import android.location.Location;
import android.location.LocationManager;

import ch.bailu.aat.preferences.Storage;
import ch.bailu.aat.preferences.location.SolidGpsTimeFix;
import ch.bailu.aat_lib.preferences.StorageInterface;
import ch.bailu.aat_lib.service.location.LocationInformation;
import ch.bailu.aat_lib.service.location.LocationStackItem;

public final class GpsLocation extends RealLocation {

    private boolean fixTime;


    public GpsLocation(LocationStackItem i, Context c, int interval) {
        super(i, c, LocationManager.GPS_PROVIDER, interval);

        fixTime = new SolidGpsTimeFix(new Storage(c)).getValue();
    }


    @Override
    protected LocationInformation factoryLocationInformation(Location location, int state) {

        LocationInformation l = new GpsLocationInformation(location, state);

        fixGpsTime(location, l.getCreationTime());

        return l;

    }

    @Override
    public void onPreferencesChanged(StorageInterface storage, String key, int presetIndex) {
        fixTime = new SolidGpsTimeFix(storage).getValue();
    }

    private void fixGpsTime(Location l, long systemTime) {
        long time = SolidGpsTimeFix.fix(l.getTime(), systemTime);

        if (fixTime) {

            l.setTime(time);
        }
    }
}

File: aat-android/src/main/java/ch/bailu/aat/services/location/GpsLocation.kt
package ch.bailu.aat.services.location

import android.content.Context
import android.location.Location
import android.location.LocationManager
import ch.bailu.aat.preferences.Storage
import ch.bailu.aat.preferences.location.SolidGpsTimeFix
import ch.bailu.aat.preferences.location.SolidGpsTimeFix.Companion.fix
import ch.bailu.aat_lib.preferences.StorageInterface
import ch.bailu.aat_lib.service.location.LocationInformation
import ch.bailu.aat_lib.service.location.LocationStackItem

class GpsLocation(i: LocationStackItem?, context: Context, interval: Int) :
    RealLocation(i, context, LocationManager.GPS_PROVIDER, interval) {
    private var fixTime: Boolean = SolidGpsTimeFix(Storage(context)).value


    override fun factoryLocationInformation(location: Location, state: Int): LocationInformation {
        val result: LocationInformation = GpsLocationInformation(location, state)
        fixGpsTime(location, result.creationTime)
        return result
    }

    override fun onPreferencesChanged(storage: StorageInterface, key: String, presetIndex: Int) {
        fixTime = SolidGpsTimeFix(storage).value
    }

    private fun fixGpsTime(location: Location, systemTime: Long) {
        val time = fix(location.time, systemTime)
        if (fixTime) {
            location.time = time
        }
    }
}
